src/Spice86.Core/Emulator/InterruptHandlers/Input/Keyboard/KeyboardInt16Handler.cs

```diff
@@ -60,80 +55,92 @@ private void Unsupported(int operation) {
-    ///     - 0xA1 => CallbackDequeueAndSetAx
-    ///     - 0xA4 => Run (default C# dispatcher for unsupported AH values)
-    ///     Notes:
-    ///     - The short jumps are encoded to skip over the DefaultDispatch+IRET and to loop back while waiting.
-    ///     - This keeps the common AH=00h path fast in RAM while preserving behavior for other functions.
+    /// Notes:
+    /// - Interrupts remain enabled during the wait loop so IRQ1 can populate the BIOS buffer.
+    /// - A short JZ right after the CMP skips exactly the default-dispatch INT+IRET sequence.
     /// ]]></remarks>
     /// <param name="memoryAsmWriter">Helper used to write machine code and callbacks into guest memory.</param>
     /// <returns>The segment:offset address where the handler stub was emitted.</returns>
     public override SegmentedAddress WriteAssemblyInRam(MemoryAsmWriter memoryAsmWriter) {
-        // Only AH=00 (GetKeystroke) is implemented in the in-RAM handler.
-        // All other AH values jump to the default C# dispatcher via callback.
-
         SegmentedAddress handlerAddress = memoryAsmWriter.CurrentAddress;
 
+        // Ensure IF is set on entry (like PC XT BIOS does)
+        memoryAsmWriter.WriteSti();
+
         // CMP AH, 00
         memoryAsmWriter.WriteUInt8(0x80);
         memoryAsmWriter.WriteUInt8(0xFC);
         memoryAsmWriter.WriteUInt8(0x00);
-        // JZ L_AH00 (+0x05 to skip the default callback and IRET)
-        memoryAsmWriter.WriteJz(4 + 1);
+        // JZ L_AH00 (+3 to skip: INT <cb>, IRET)
+        memoryAsmWriter.WriteJz(3);
 
-        // L_DEFAULT: callback DefaultDispatch then IRET
+        // L_DEFAULT: callback DefaultDispatch, then IRET
         memoryAsmWriter.RegisterAndWriteCallback(VectorNumber, Run);
         memoryAsmWriter.WriteIret();
 
-        // L_AH00:
-        // callback HasKey (sets ZF=0 when key present)
-        memoryAsmWriter.RegisterAndWriteCallback(CallbackHasKey);
-        // JNZ have_key (+0x04 to skip INT 09h and the backward jump)
-        memoryAsmWriter.WriteJnz(4);
-        // INT 09h
-        memoryAsmWriter.WriteInt(0x09);
-        // JMP short back to L_AH00 (-0x10)
-        memoryAsmWriter.WriteJumpShort(-10);
-        // have_key: dequeue and set AX, then IRET
+        // L_AH00: Busy-wait using GetKeystrokeStatus callback (sets AX and ZF)
+        SegmentedAddress ah00Address = memoryAsmWriter.CurrentAddress;
+
+        // Call C# status: ZF=1 if empty (AX=0), ZF=0 and AX=key if pending (non-destructive)
+        memoryAsmWriter.RegisterAndWriteCallback(CallbackGetKeystrokeStatus);
+
+        // If key is available (ZF=0), skip the back-jump; else loop
+        memoryAsmWriter.WriteJnz(2);
+
+        // JMP short L_AH00
+        sbyte offset = (sbyte)(ah00Address.Offset - (memoryAsmWriter.CurrentAddress.Offset + 2));
+        memoryAsmWriter.WriteJumpShort(offset);
+
+        // Key is available: dequeue atomically and return
+        memoryAsmWriter.WriteCli();
```

Comment:
Do we need this? When callback is executing interrupts are not checked, this is pure C#